### ISSUE TITLE
Panic on rke up

### ISSUE THREAD

**Issue by stiks** [rancher/rke#867]
I'm having issues with every cluster I run after 2.0.4. Time to time etcd start eating all cpu/memmory then crash. I cannot trace issue, i believe there's some kind of memory leak.

This issue happened when I've tried to rescue my cluster rancher cluster. Control place node died, and I cannot restore it, I have to replace control plane in kube_config file to another control plane.

**RKE version:**
rke version v0.1.9

**Operating system and kernel: (`cat /etc/os-release`, `uname -r` preferred)**
MacOS

**Type/provider of hosts: (VirtualBox/Bare-metal/AWS/GCE/DO)**
Bare-metal

**Steps to Reproduce:**
I've no idea how to make it reproducible

**Results:**
```
INFO[0005] [state] Successfully Fetched cluster state to Kubernetes ConfigMap: cluster-state
INFO[0005] [certificates] Getting Cluster certificates from Kubernetes
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 1 [running]:
github.com/rancher/rke/vendor/github.com/urfave/cli.HandleAction.func1(0xc420594700)
	/go/src/github.com/rancher/rke/vendor/github.com/urfave/cli/app.go:472 +0x28d
panic(0x1f65c40, 0x2a04fe0)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/rancher/rke/cluster.getClusterCerts(0x29c88c0, 0xc420014840, 0xc420384960, 0xc4202b1b40, 0x3, 0x4, 0x0, 0x1dd16f2, 0xc4200a4000)
	/go/src/github.com/rancher/rke/cluster/certificates.go:152 +0x877
github.com/rancher/rke/cluster.(*Cluster).GetClusterState(0xc4200a4000, 0x29c88c0, 0xc420014840, 0x7ffeefbffa00, 0x0, 0x0)
	/go/src/github.com/rancher/rke/cluster/state.go:71 +0x440
github.com/rancher/rke/cmd.ClusterUp(0x29c88c0, 0xc420014840, 0xc420410e00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/rancher/rke/cmd/up.go:84 +0x191
github.com/rancher/rke/cmd.clusterUpFromCli(0xc4204588c0, 0x0, 0x0)
	/go/src/github.com/rancher/rke/cmd/up.go:200 +0x365
reflect.Value.call(0x1f01420, 0x21a1100, 0x13, 0x210a7d1, 0x4, 0xc4202f26a0, 0x1, 0x1, 0xc4202f2628, 0x20d1c60, ...)
	/usr/local/go/src/reflect/value.go:434 +0x91f
reflect.Value.Call(0x1f01420, 0x21a1100, 0x13, 0xc4202f26a0, 0x1, 0x1, 0x140, 0xc4204588c0, 0x0)
	/usr/local/go/src/reflect/value.go:302 +0xa4
github.com/rancher/rke/vendor/github.com/urfave/cli.HandleAction(0x1f01420, 0x21a1100, 0xc4204588c0, 0x0, 0x0)
	/go/src/github.com/rancher/rke/vendor/github.com/urfave/cli/app.go:481 +0x198
github.com/rancher/rke/vendor/github.com/urfave/cli.Command.Run(0x210a210, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x211f98c, 0x14, 0x0, ...)
	/go/src/github.com/rancher/rke/vendor/github.com/urfave/cli/command.go:186 +0xab8
github.com/rancher/rke/vendor/github.com/urfave/cli.(*App).Run(0xc420466000, 0xc42000e1c0, 0x4, 0x4, 0x0, 0x0)
	/go/src/github.com/rancher/rke/vendor/github.com/urfave/cli/app.go:235 +0x5d6
main.mainErr(0xc4202f3f78, 0x1dfbbd8)
	/go/src/github.com/rancher/rke/main.go:52 +0x12d3
main.main()
	/go/src/github.com/rancher/rke/main.go:16 +0x22
```

**Comment by mpsnp** [comment]
Also faced this issue

**Comment by stiks** [comment]
This is my best guess

I have 5 etcd (3 of them with control plane) ~ 2GB 2 CPU baremetal instances
When network gets a bit unstable ectd split brained. Because 2GB is not really enough for etcd+control plane to get itself sorted, scheduler trying to move pods around and check live probles. When instance run our of memory, etcd and control pane crashes and keep restarting as well as canal. With crashed canal there's nothing you can do apart from restarting everything. In my case I've shutdown couple control planes to get traffic under control.

As result, when I do `rke up`, it start thinking that this is brand new setup and generates certificate. And can actually finish sucessfully, but I woudn't be able to add this nodes back.

**Comment by galal-hussein** [comment]
@stiks running rke up on a failed cluster may yield unexpected results because rke tries to communicate with the k8s cluster to get the state and certificates saved there, i am not able to reproduce on my setup though, can you try to get more information from your setup including:
```
> docker logs etcd (on each etcd node)
> docker ps -a
if you have access to k8s
> kubectl get configmaps -n kube-system
> kubectl get secrets -n kube-system
```

**Comment by sangeethah** [comment]
There was 1 time were I could reproduce the panic when the  fix was still not made available:

Had a DO cluster with 1 node(all roles) provisioned successfully.
After I increased the scale of the nodepool to 3 , cluster provisioning is stuck in "updating" forever with status- Failed to get kubernetes node [santestnew1]: "santestnew1"
Rancher sever crashed when rke is updating the cluster.
 
```2018/09/05 10:48:43 [INFO] Provisioning cluster [c-z2pb9]
2018/09/05 10:48:43 [INFO] Updating cluster [c-z2pb9]
2018/09/05 10:48:43 [INFO] cluster [c-z2pb9] provisioning: Building Kubernetes cluster
2018/09/05 10:48:44 [INFO] cluster [c-z2pb9] provisioning: [dialer] Setup tunnel for host [165.227.125.107]
2018/09/05 10:48:44 [INFO] cluster [c-z2pb9] provisioning: [dialer] Setup tunnel for host [159.89.183.188]
2018/09/05 10:48:44 [INFO] cluster [c-z2pb9] provisioning: [dialer] Setup tunnel for host [159.89.183.176]
2018/09/05 10:48:44 [INFO] cluster [c-z2pb9] provisioning: [state] Found local kube config file, trying to get state from cluster
2018/09/05 10:48:44 [INFO] cluster [c-z2pb9] provisioning: [state] Fetching cluster state from Kubernetes
2018/09/05 10:48:44 [INFO] cluster [c-z2pb9] provisioning: [state] Successfully Fetched cluster state to Kubernetes ConfigMap: cluster-state
2018/09/05 10:48:44 [INFO] cluster [c-z2pb9] provisioning: [certificates] Getting Cluster certificates from Kubernetes
panic: runtime error: index out of range

goroutine 30628711 [running]:
github.com/rancher/rancher/vendor/github.com/rancher/rke/cluster.getClusterCerts(0xc133ce0, 0xc442e78510, 0xc44a06aff0, 0xc42da629e0, 0x3, 0x4, 0x0, 0xc441efc000, 0x0)
    /go/src/github.com/rancher/rancher/vendor/github.com/rancher/rke/cluster/certificates.go:152 +0xab7
github.com/rancher/rancher/vendor/github.com/rancher/rke/cluster.(*Cluster).GetClusterState(0xc441efc000, 0xc133ce0, 0xc442e78510, 0x0, 0x0, 0x0)
    /go/src/github.com/rancher/rancher/vendor/github.com/rancher/rke/cluster/state.go:71 +0x42f
github.com/rancher/rancher/vendor/github.com/rancher/rke/cmd.ClusterUp(0xc133ce0, 0xc442e78510, 0xc43be28300, 0xc4277abdf0, 0xc4277abe00, 0xc4403c3d00, 0x0, 0xc440cdfa10, 0x2e, 0x0, ...)
    /go/src/github.com/rancher/rancher/vendor/github.com/rancher/rke/cmd/up.go:79 +0x191
github.com/rancher/rancher/vendor/github.com/rancher/kontainer-engine/drivers/rke.(*Driver).Update(0xc42082b950, 0xc133ce0, 0xc442e78510, 0xc44080a790, 0xc4403c3320, 0x0, 0x0, 0x0)
    /go/src/github.com/rancher/rancher/vendor/github.com/rancher/kontainer-engine/drivers/rke/rke_driver.go:182 +0x24c
github.com/rancher/rancher/vendor/github.com/rancher/kontainer-engine/types.(*GrpcServer).Update(0xc420881600, 0xc133ce0, 0xc442e78300, 0xc44864f010, 0xc420881600, 0xc4356a3740, 0xc4317ef990)
    /go/src/github.com/rancher/rancher/vendor/github.com/rancher/kontainer-engine/types/rpc_server.go:55 +0x79
github.com/rancher/rancher/vendor/github.com/rancher/kontainer-engine/types._Driver_Update_Handler(0x510b260, 0xc420881600, 0xc133ce0, 0xc442e78300, 0xc420121c70, 0x0, 0x0, 0x0, 0x434054, 0xc449fe5380)
    /go/src/github.com/rancher/rancher/vendor/github.com/rancher/kontainer-engine/types/drivers.pb.go:598 +0x276
github.com/rancher/rancher/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc420f48b00, 0xc1482c0, 0xc438eab500, 0xc4254c92c0, 0xc421bc5c80, 0xc0d1cf8, 0x0, 0x0, 0x0)
    /go/src/github.com/rancher/rancher/vendor/google.golang.org/grpc/server.go:921 +0x96e
```

I have not seen this issue when testing with rancher-server version that had this rke fix.

Also there have been different performance issues relating to memort leaks which has been addressed in 2.0.8 release - https://github.com/rancher/rancher/issues/14356.
